Objects/typeobject.c

```diff
@@ -6190,15 +6190,17 @@ object_set_class(PyObject *self, PyObject *value, void *closure)
         /* Changing the class will change the implicit dict keys,
          * so we must materialize the dictionary first. */
         if (oldto->tp_flags & Py_TPFLAGS_INLINE_VALUES) {
-            PyDictObject *dict = _PyObject_ManagedDictPointer(self)->dict;
-            if (dict == NULL) {
-                dict = (PyDictObject *)_PyObject_MakeDictFromInstanceAttributes(self);
-                if (dict == NULL) {
-                    return -1;
-                }
-                _PyObject_ManagedDictPointer(self)->dict = dict;
+            PyDictObject *dict = _PyObject_MaterializeManagedDict(self);
+            bool error = false;
+
+            Py_BEGIN_CRITICAL_SECTION2(self, dict);
+
+            if (dict == NULL || _PyDict_DetachFromObject(dict, self) < 0) {
+                error = true;
             }
-            if (_PyDict_DetachFromObject(dict, self)) {
+
+            Py_END_CRITICAL_SECTION2();
```

Comment:
The `dict == NULL` check has to be outside of the critical section. `Py_BEGIN_CRITICAL_SECTION2` does not handle `NULL` arguments.